### ISSUE TITLE
[Proposal] Create an adapter to use PubSub components as Bindings

### ISSUE THREAD

**Issue by pkedy** [dapr/components-contrib#1305]
Currently, there are several bindings that have a pubsub counterpart (E.g. Kafka, Rabbit, Pulsar, NATS, Redis). In many cases the code in the PubSub component is far more robust. Given the similarity between bindings (Data + Metadata) and pubsub (Topic + Data + Metadata), an adapter could be used that specifies the read/invoke topics to subscribe and publish to. The primary motivation is that adapter reuses code and users can expect the same behavior between the binding and pubsub.

Here are some issues that would possibly be resolved for Kafka:

#1277
#1084
#955
#1006
#168

The downside is that if there are current inconsistencies in the component settings, then migrating to this approach would break existing versions. Since none of these components are marked as stable (yet), now is the time to do this if at all.

**Comment by daixiang0** [comment]
It is a design issue as bindings are more abstract than pubsub, we should better guide users to choose the most suitable one instead of making the two more obscure.

**Comment by pkedy** [comment]
> It is a design issue as bindings are more abstract than pubsub, we should better guide users to choose the most suitable one instead of making the two more obscure.

Hi, @daixiang0! I don't disagree with this statement. I have opinions about this duality as well; however, users are _already_ using the binding versions of pubsub components and experiencing issues that are not present in the pubsub component. This proposal isn't intended to debate the merit of bindings for message brokers. Instead, this proposal aims to solve is the fact that there is similar code that in some cases is not as robust as the code available in pubsub. If anything it provides a simpler, more maintainable option for developers to subscribe/publish to topics without having to configure or have knowledge of pubsub/topics inside their applications.

**Comment by sthussey** [comment]
Might this also be improved by simply refactoring pubsub and binding components to use shared code like the Redis components? That would allow some foundational aspects like service connectivity, authentication, etc... to be used in both even if the functionality diverges.

**Comment by skyao** [comment]
I have a question about the design principle of pubsub and binding: 

since we have pubsub components for these message broker, why we still provide binding components for them? 

Do we provide more capabilities than pubsub in binding for these message broker?

**Comment by artursouza** [comment]
@skyao I agree. I think we should remove Kafka input binding.

**Comment by yaron2** [comment]
We should not remove the bindings that have pub/sub equivalents, specifically Kafka, because while they allow users to contribute and use more specialized features from the underlying system. For example, a Kafka binding can expose an API that returns the number of messages in a topic.

The fact it's not there today does not mean we should block this functionality from being added at any point in the future by removing the component.

**Comment by artursouza** [comment]
That is a great point, @yaron2. Does it mean we should not simply make the Kafka binding a wrap of the pubsub component and keep those as completely separate implementations?

**Comment by yaron2** [comment]
We can use the suggested adopter pattern to reuse what code we can (if at all, but I believe most code can be reused).

As long as we allow bindings to add operations (which is their goal for flexibility) and their own unique metadata, we can refactor the code however we see fit.

**Comment by skyao** [comment]
@artursouza  @yaron2  @pkedy I have done the code refactory for kafka pubsub component and kafka binding component accroding to above discuss. 

I split this code refactory into two parts:

1. refactory kafka pubsub compoent to extract the common kafka code

     These common kafka code is moved to `internal/kafka`  .

     Kafka pubsub compoent now is based on this common code and it adapters its pubsub api method to the common code like init() / publish() / subscribe() / close(). 

    This refactory should be safe since it only split and move the code, no functionality change.

2. refactory kafka binding component to reuse the common kafka code

    This is a BIG change to kafka binding component since now it's code base is moved to the common kafka code extracting from kafka pubsub component. 

    The benefit is that now kafka component is have the same features and implementations as kafka pubsub component. 

    The challenge is that we need to careful handle the difference.

The second PR should be reviewed very carefully.

**Comment by artursouza** [comment]
I suggest we keep the adapted version as kafka binding v2 and keep the old version as-is in v1. This way, we don't break existing apps. Even as alpha component, kafka is a popular broker.

**Comment by yaron2** [comment]
Do we know of breaking changes here?

**Comment by artursouza** [comment]
> Do we know of breaking changes here?

Yes. The E2E tests are failing already because of that in the PR https://github.com/dapr/dapr/actions/runs/2243065335
